File: job-2022-autumn/1038.binary-search-tree-to-greater-sum-tree.js
/*
 * @lc app=leetcode id=1038 lang=javascript
 *
 * [1038] Binary Search Tree to Greater Sum Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var bstToGst = function (root) {
    helper(root, 0);
    return root;
};

/**
 * 递归降序遍历和
 * @param {TreeNode|null} root
 * @param {number} currentSum - 在以 root 为根的子树之外，比 root 更大的结点的和
 * @returns {number} - 本子树的结点的和
 */
function helper(root, currentSum) {
    if (root === null) return currentSum;

    currentSum = helper(root.right, currentSum);
    currentSum += root.val;
    root.val = currentSum;
    currentSum = helper(root.left, currentSum);

    return currentSum;
}
// @lc code=end

File: prev-2024/1038.binary-search-tree-to-greater-sum-tree.cpp
/*
 * @lc app=leetcode id=1038 lang=cpp
 *
 * [1038] Binary Search Tree to Greater Sum Tree
 */
struct TreeNode {
  int val;
  TreeNode* left;
  TreeNode* right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode* left, TreeNode* right)
      : val(x), left(left), right(right) {}
};
// @lc code=start
class Solution {
 public:
  Solution() { this->currentSum = 0; }

  TreeNode* bstToGst(TreeNode* root) {
    this->inOrderTraverse(root);
    return root;
  }

 private:
  int currentSum;

  void inOrderTraverse(TreeNode* root) {
    if (root == nullptr) {
      return;
    }
    this->inOrderTraverse(root->right);
    currentSum += root->val;
    root->val = currentSum;
    this->inOrderTraverse(root->left);
  }
};
// @lc code=end
